### PR TITLE
Make complete packages mockable

### PR DESCRIPTION
**Pull request by as-design**
Hi there, we have a lot data classes in our project that we have to mock. Instead of making each class mockable it would be handy to make complete packages mockable. Here is a proposal.

### CODE REVIEW

**Comment by joke** [comment]
The changes did not built successfully. I created a new pull requests #136 with a few additions.

Thank you very much. I will release a new version.

**Comment by joke** [comment]
Okay change of plans :)

File: examples/kotlin/src/test/groovy/io/github/joke/spockmockable/PersonTest.groovy [hunk 1]
```diff
@@ -3,7 +3,7 @@ package io.github.joke.spockmockable
 import org.spockframework.mock.MockUtil
 import spock.lang.Specification
 
-@Mockable([Person, Address])
+@Mockable(packages = "io.github.joke.spockmockable")
```
Comment (joke) [inline examples/kotlin/src/test/groovy/io/github/joke/spockmockable/PersonTest.groovy:6]: Could you add a separate test file. So we don't loose a test testing lists classes.

File: spock-mockable/src/main/groovy/io/github/joke/spockmockable/internal/MockableASTTransformation.groovy [hunk 2]
```diff
@@ -18,12 +18,14 @@ import static org.codehaus.groovy.control.CompilePhase.OUTPUT
...
 
     private static Set<String> detectedClasses = [] as Set<String>
+    private static Set<String> detectedPackages = [] as Set<String>
     private static CompilationUnit compilationUnit
 
     @Override
     void visit(ASTNode[] nodes, SourceUnit source) {
         def annotation = nodes[0] as AnnotationNode
         detectedClasses.addAll(getClassList(annotation, 'value')*.name as List)
+        detectedPackages.addAll(getMemberStringList(annotation, 'packages')?:[])
```
Comment (joke) [inline spock-mockable/src/main/groovy/io/github/joke/spockmockable/internal/MockableASTTransformation.groovy:28]: needs to be `getMemberList` for groovy 2.5 backwards compatibility

File: spock-mockable/src/main/java/io/github/joke/spockmockable/internal/MockableExtension.java [hunk 3]
```diff
@@ -64,35 +66,61 @@ private static void installTransformer() {
...
+    private static boolean isNotATest(TypeDescription typeDescription) {
+        return !typeDescription.isAssignableTo(Specification.class);
+    }
+
+    private static boolean isInClasses(Set<String> classes, TypeDescription typeDescription) {
+        return classes.contains(typeDescription.getName());
+    }
+
+    private static boolean isInPackages(Set<String> packages, TypeDescription typeDescription) {
+        return packages.stream().anyMatch(packageName -> Optional.ofNullable(typeDescription.getCanonicalName()).map(canonicalName -> canonicalName.startsWith(packageName + ".")).orElse(false));
```
Comment (joke) [inline spock-mockable/src/main/java/io/github/joke/spockmockable/internal/MockableExtension.java:121]: split the stream to make it a little more readable.

```
.stream()
.anyMatch()
.map()
.orElse()
```

**Review by joke — APPROVED** [review]

**Comment by joke** [comment]
please squash the commits and start the commit message with 'feat: Make complete...'

**Review by joke — CHANGES_REQUESTED** [review]
please squash the commits and start the commit message with 'feat: Make complete...'

File: examples/kotlin/src/test/groovy/io/github/joke/spockmockable/PersonTest.groovy (hunk 1 above)
Comment (as-design) [reply examples/kotlin/src/test/groovy/io/github/joke/spockmockable/PersonTest.groovy]: Of course. I added seperate tests for this feature.

File: spock-mockable/src/main/groovy/io/github/joke/spockmockable/internal/MockableASTTransformation.groovy (hunk 2 above)
Comment (as-design) [reply spock-mockable/src/main/groovy/io/github/joke/spockmockable/internal/MockableASTTransformation.groovy]: Make sense. I fixed it.

File: spock-mockable/src/main/java/io/github/joke/spockmockable/internal/MockableExtension.java (hunk 3 above)
Comment (as-design) [reply spock-mockable/src/main/java/io/github/joke/spockmockable/internal/MockableExtension.java]: Of course.

**Comment by as-design** [comment]
Thanks for the review! I am done with the changes. Also added a bit to the docs. Looking forward to the release.

**Review by joke — CHANGES_REQUESTED** [review]

File: examples/complete-packages/build.gradle [hunk 4]
```diff
@@ -0,0 +1,18 @@
+plugins {
+    id 'java'
+    id 'groovy'
+    id 'org.jetbrains.kotlin.jvm' version '1.6.10'
+}
+
+dependencies {
+    testImplementation platform('org.codehaus.groovy:groovy-bom:3.0.9')
```
[the remark this answers is not in the document]
Comment (as-design) [reply examples/complete-packages/build.gradle]: Alright. I added the tests in the existing examples.

**Review by joke — APPROVED** [review]

**Pull request merged**